File: library/src/main/ts/libraryBrs/src/app/core/services/rest/borrowings.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable } from 'rxjs';
import { Borrowing, Response } from './api/api';
import { RestService } from './rest.service';
const URL = "api/borrowings";

@Injectable()
export class BorrowingsService {
  public borrowings: BehaviorSubject<Borrowing[]> = new BehaviorSubject<Borrowing[]>([]);
  public pastBorrowings: BehaviorSubject<Borrowing[]> = new BehaviorSubject<Borrowing[]>([]);
  

  constructor(private restService: RestService) { }

  getByUserId(userId:string): any {
    return this.restService.get<Response<Borrowing[]>>(`${URL}/user/${userId}`).subscribe((resp:Response<Borrowing[]>) => {
      if(resp.content){
         this.borrowings.next(resp.content);
      }
     
    })
  }

  getPastByUserId(userId: string): any {
    return this.restService.get<Response<Borrowing[]>>(`${URL}/user/${userId}/past`).subscribe((resp:Response<Borrowing[]>) => {
      if (resp.content) {
        this.pastBorrowings.next(resp.content);
      }
    })
  }

  prolong(id:string): Observable<Response<any>> {
    return this.restService.patch(`${URL}/${id}/prolong`, {});
  }

  create(borrowings: Borrowing[]): Observable<Response<any>> {
    return this.restService.post(`${URL}`, borrowings);
  }

  returnBorrowing(bookCopiesIds:string[]): Observable<Response<any>> {
    let stringIds="";
    bookCopiesIds.forEach(id=>stringIds+`&bookCopiesIds=${id}`);
    stringIds=stringIds.substring(1);
    return this.restService.patch(`${URL}/return?${stringIds}`, {});
  }
}

File: library/src/main/dart/mobile/librarian/lib/services/borrowingsService.dart
import 'dart:collection';

import 'package:flutter/cupertino.dart';
import 'package:librarian/services/api/api.dart';
import 'package:librarian/services/restService.dart';

class BorrowingsService extends ChangeNotifier {
  final String _url = "api/borrowings";
  final List<Borrowing> _borrowings = [];
  UnmodifiableListView<Borrowing> get items =>
      UnmodifiableListView(_borrowings);

  Borrowing? _borrowing;
  Borrowing? get item => (_borrowing);

  void create( List<Borrowing> borrowings, void Function(dynamic) onSuccess) {
    RestService rest = RestService();
    List<Map<dynamic, dynamic>> borrowingsJson = borrowings.map((borrowing) => borrowing.toJson()).toList();
    rest.post<dynamic>(
        path: _url, body: borrowingsJson, onSuccess: onSuccess, onError: (j)=>print(j));
  }

  void createBorrowing(List<BookCopy> bookCopies, AppUser user, void Function(dynamic) onSuccess){
    List<Borrowing> borrowings = bookCopies.map((copy) => 
    Borrowing(null, copy, copy.id, user, user.id, null, null, null, null)).toList();
    create(borrowings, onSuccess);

  }

  void onSuccessCreate(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void getByUserId(String userId) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/user/${userId}", onSuccess: onSuccessGetByUserId);
  }

  void onSuccessGetByUserId(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void getPastByUserId(String userId) {
    RestService rest = RestService();
    rest.get<dynamic>(
        path: "${_url}/user/${userId}/past",
        onSuccess: onSuccessGetPastByUserId);
  }

  void onSuccessGetPastByUserId(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }

  void returnBorrowing(List<BookCopy> bookCopies, void Function(dynamic) onSuccess) {
    RestService rest = RestService();
    String stringIds = "";
    bookCopies.forEach((bookCopy) => stringIds += "&bookCopiesIds=${bookCopy.id}");
    stringIds = stringIds.substring(0, stringIds.length - 1);
    print(stringIds);
    rest.patch<dynamic>(
        path: "${_url}/return?${stringIds}",
        onSuccess: onSuccess,
        onError: (j)=>print(j));
  }

  void onSuccessReturnBorrowing(dynamic object) {
    print(object);
    Response response = Response.fromJson(object);
    notifyListeners();
  }
}
